File: lib/src/features/admin/skilltrees/domain/skilltree.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:hero/src/features/characters/domain/character.dart';
import 'package:hero/src/features/characters/domain/character_overview.dart';

import 'node.dart';

part 'skilltree.freezed.dart';
part 'skilltree.g.dart';

@freezed
class Skilltree with _$Skilltree {
  const factory Skilltree({
    @Default("") String id,
    @Default("") String name,
    CharacterOverview? character,
    @Default(0) int points,
    @Default(true) bool isActiveTree,
    @Default([]) List<Node> nodes,
  }) = _Skilltree;

  factory Skilltree.fromJson(Map<String, dynamic> json) => _$SkilltreeFromJson(json);
}

extension SkilltreeExtension on Skilltree {
  Skilltree updateNodes(List<String> idsToDelete, List<Node> newNodes) {
    return copyWith(nodes: [...nodes.where((element) => !idsToDelete.contains(element.id)), ...newNodes]);
  }

  bool isNodeUnlockable(String nodeId) {
    Node nodeToCheck = nodes.singleWhere((element) => nodeId == element.id);

    return nodeToCheck.isEasyReachable
        ? nodes.where((element) => nodeToCheck.precessors.contains(element.id)).any((element) => element.isUnlocked)
        : !nodes.where((element) => nodeToCheck.precessors.contains(element.id)).any((element) => !element.isUnlocked);
  }
}

File: Server/Hero.Server.Core/Models/Skilltree.cs
namespace Hero.Server.Core.Models
{
    public class Skilltree
    {
        public Guid Id { get; set; }
        public Guid GroupId { get; set; }
        public Guid? CharacterId { get; set; }
        public string Name { get; set; }
        public bool IsActiveTree { get; set; }
        public int Points { get; set; }
        public List<SkilltreeNode> Nodes { get; set; }
        public Character? Character { get; set; }

        public void Update(Skilltree tree)
        {
            this.Name = tree.Name;
            this.Points = tree.Points;
            this.IsActiveTree = tree.IsActiveTree;
            this.Nodes = tree.Nodes;
        }
    }
}
